src/main/java/org/aksw/iguana/cc/worker/impl/SPARQLProtocolWorker.java

```diff
@@ -307,128 +283,211 @@ private ExecutionStats executeQuery(Duration timeout, boolean discardOnFailure)
+            try {
+                request = requestFactory.buildHttpRequest(
+                        queryHandle,
+                        config().connection(),
+                        config().acceptHeader()
+                );
+            } catch (IOException | URISyntaxException e) {
+                return createFailedResultBeforeRequest(queryHandle.index(), e);
+            }
 
-        try {
-            request = requestFactory.buildHttpRequest(
-                    queryHandle.queryInputStream(),
-                    timeout,
-                    config().connection(),
-                    config().acceptHeader()
-            );
-        } catch (IOException | URISyntaxException e) {
-            return new HttpExecutionResult(
-                    queryHandle.index(),
-                    Optional.empty(),
-                    Instant.now(),
-                    Duration.ZERO,
-                    Optional.empty(),
-                    OptionalLong.empty(),
-                    OptionalLong.empty(),
-                    Optional.of(e)
-            );
+            // set queryIndex to the index of the queryHandle, so that the result can be associated with the query
+            queryIndex = queryHandle.index();
         }
 
-        // check if the last execution task is stuck
-        if (this.httpClient.executor().isPresent() && ((ThreadPoolExecutor) this.httpClient.executor().get()).getActiveCount() != 0) {
-            // This might never cancel the task if the client that's connected to is broken. There also seems to be a
-            // bug where the httpClient never properly handles the interrupt from the shutdownNow method.
-            // See: https://bugs.openjdk.org/browse/JDK-8294047
-            ((ThreadPoolExecutor) this.httpClient.executor().get()).shutdownNow();
-            final var waitStart = Instant.now();
-            try {
-                while (!((ThreadPoolExecutor) this.httpClient.executor().get()).awaitTermination(1, TimeUnit.SECONDS)) {
-                    LOGGER.warn("{}\t:: [Thread-ID: {}]\t:: Waiting for the http client to shutdown. Elapsed time: {}", this, Thread.currentThread().getId(), Duration.between(waitStart, Instant.now()));
+        // execute the request
+        final Instant timeStamp = Instant.now();
+        final var requestStart = System.nanoTime();
+        final var future = httpClient.execute(request, new AbstractBinResponseConsumer<HttpExecutionResult>() {
+
+            private HttpResponse response;
+            private final StreamingXXHash64 hasher = hasherFactory.newStreamingHash64(0);
+            private long responseSize = 0; // will be used if parseResults is false
+            private long responseEnd = 0;  // time in nanos
+
+            @Override
+            public void releaseResources() {} // nothing to release
+
+            @Override
+            protected int capacityIncrement() {
+                return Integer.MAX_VALUE; // get as much data in as possible
```

Comment:
I am not completely sure where exactly this size ends up being used. I didn't find the documentation of AbstractBinDataConsumer very helpful for this. For safety reasons I would suggest to use: `Integer.MAX_VALUE - 8` instead. That way it is guaranteed that it would not throw creating a java byte array with the capacity.